pkg/k8sclient/bastions.go

```diff
@@ -0,0 +1,114 @@
+package k8sclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/giantswarm/microerror"
+	capg "sigs.k8s.io/cluster-api-provider-gcp/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const BastionLabelKey = "cluster.x-k8s.io/deployment-name"
+
+type Bastions struct {
+	client    client.Client
+	finalizer string
+}
+
+func NewBastions(client client.Client, finalizer string) *Bastions {
+	return &Bastions{
+		client:    client,
+		finalizer: finalizer,
+	}
+}
+
+func (b *Bastions) GetBastionIPList(ctx context.Context, cluster *capg.GCPCluster) ([]string, error) {
+	machineList, err := b.getBastionMachineList(ctx, cluster)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var bastionPublicIPList []string
+
+	for _, machine := range machineList.Items {
+		// no ip address yet
+		if len(machine.Status.Addresses) == 0 {
+			return nil, errors.New("bastion IP is not yet available")
+		}
+
+		// get the public IP
+		for _, addr := range machine.Status.Addresses {
+			if addr.Type == "ExternalIP" {
+				bastionPublicIPList = append(bastionPublicIPList, addr.Address)
+				break
+			}
+		}
+	}
+
+	if len(bastionPublicIPList) == 0 {
+		return nil, errors.New("bastion IP is not yet available")
+	}
```

Comment:
wouldn't we go through this code path when no bastion node are present? I was wondering if that should be an error for the dns-operator or not.

Reply:
I could imagine a scenario when some GCPMachine with bastion label is up and the field with addresses is partially populated but public IP is not yet available

I cant vouch fro GCP serving all information for the machine at all time

if you feel this is nonsence I can remove it